[JSC] Implement core of sequestered arena allocator

Achierius opened the pull request:
#### 9dd52f49f7c30106f48751f6776a01e948ee97e0
<pre>
[JSC] Implement core of sequestered arena allocator
<a href="https://bugs.webkit.org/show_bug.cgi?id=287524">https://bugs.webkit.org/show_bug.cgi?id=287524</a>
<a href="https://rdar.apple.com/114840482">rdar://114840482</a>

Reviewed by Yusuke Suzuki.

This implements the core functionality of a new arena allocator built
around a sequestered pool of memory. The intention is for this allocator
to be used by the JIT compiler for allocating datastructures; as of
right now, only a small number of datatypes are set up to use the new
allocator. As this is an arena allocator, users need to specify the
lifetime of their allocations indirectly via the ArenaLifetime object.
Currently, the allocator is disabled, and has been stripped of several
unstable components which are still under development. In particular,
the `free` flow is intended to hook into the libpas scavenger, but right
now eagerly frees to tide us over; moreover, each thread&apos;s allocator
will eventually have at least a small number (2-3) of different
size-classed arenas, but for now we only use one.

This allocator is by design separate from libpas. This is because we
want to isolate not only the backing memory but also all allocator
metadata and internal memory. To do this within libpas would require we
duplicate essentially the entire libpas heap hierarchy, from
pas_large_heap up to the various bootstrap heaps, and color them
according to which pool of pages they&apos;re supposed to pull from.

Performance-wise, turning this on has no effect at the moment, but the
hope is that in time it will prove to be a progression. The arena
semantics allow us to batch all frees from a given compiler pass into
one, reducing the actual cost of `free` to a no-op as well as reducing
the load on the scavenger. It also allows us to significantly simplify
the allocation path, removing some overhead even in the shortest-path
malloc case and making it possible to stick to the bump-allocation
regime without ever having to fall back on free lists or &amp;c.

In time, we will need to add a persistent allocator which allocates from
the same pool as well -- something like SEQUESTERED_PERSISTENT_MALLOC.

With regards to the naming, I considered a couple of alternatives:
  - Protected JIT Arena Malloc
  - JIT Arena Malloc
  - Secure Arena Malloc
For the first two, I want to avoid using the term &quot;JIT&quot; as it&apos;s very
overloaded -- I&apos;ve seen it refer to things relevant to
  A) The JIT compiler
  B) Allocating RWX memory
  C) Threads executing in previously JITted code
and so would like to avoid using it here. It also has the problem that
the short-form (JaMalloc) sounds a lot like jemalloc, which could cause
further confusion.
&quot;Secure&quot; would imply that libpas/FastMalloc are &quot;insecure&quot;, which I also
want to avoid.
&quot;Sequestered&quot; is nice because it describes exactly what characterizes
this allocator: it uses memory sequestered from the rest of the system,
and it allocates it using an arena.
Thus SequesteredArenaMalloc.

* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/jit/JIT.cpp:
(JSC::JIT::~JIT):
* Source/JavaScriptCore/jit/JIT.h:
* Source/JavaScriptCore/jit/JITAllocator.h:
* Source/JavaScriptCore/jit/JITWorklistThread.cpp:
* Source/JavaScriptCore/jit/SecureAllocator.cpp: Copied from Source/JavaScriptCore/jit/JITAllocator.h.
* Source/JavaScriptCore/jit/SecureAllocator.h: Copied from Source/JavaScriptCore/jit/JITAllocator.h.
* Source/JavaScriptCore/jit/SecureAllocatorInlines.h: Copied from Source/JavaScriptCore/jit/JITAllocator.h.
* Source/JavaScriptCore/runtime/InitializeThreading.cpp:
(JSC::initialize):
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/Bag.h:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/CheckedRef.h:
(WTF::CanMakeCheckedPtr::~CanMakeCheckedPtr):
(WTF::CanMakeThreadSafeCheckedPtr::~CanMakeThreadSafeCheckedPtr):
* Source/WTF/wtf/EmbeddedFixedVector.h:
* Source/WTF/wtf/FastMalloc.cpp:
(WTF::tryFastMalloc):
(WTF::fastMalloc):
(WTF::tryFastCalloc):
(WTF::fastCalloc):
(WTF::fastRealloc):
(WTF::tryFastRealloc):
(WTF::fastZeroedMalloc):
(WTF::tryFastZeroedMalloc):
(WTF::fastAlignedMalloc):
(WTF::tryFastAlignedMalloc):
(WTF::fastCompactMalloc):
(WTF::fastCompactZeroedMalloc):
(WTF::tryFastCompactZeroedMalloc):
(WTF::fastCompactRealloc):
(WTF::fastCompactAlignedMalloc):
(WTF::tryFastCompactAlignedMalloc):
(WTF::tryFastCompactMalloc):
(WTF::tryFastCompactRealloc):
(WTF::ForbidMallocUseForCurrentThreadScope::ForbidMallocUseForCurrentThreadScope): Deleted.
(WTF::ForbidMallocUseForCurrentThreadScope::~ForbidMallocUseForCurrentThreadScope): Deleted.
(WTF::DisableMallocRestrictionsForCurrentThreadScope::DisableMallocRestrictionsForCurrentThreadScope): Deleted.
(WTF::DisableMallocRestrictionsForCurrentThreadScope::~DisableMallocRestrictionsForCurrentThreadScope): Deleted.
* Source/WTF/wtf/FastMalloc.h:
(WTF::FastMalloc::fastFree):
(WTF::FastCompactMalloc::fastFree):
(WTF::ForbidMallocUseForCurrentThreadScope::~ForbidMallocUseForCurrentThreadScope): Deleted.
(WTF::DisableMallocRestrictionsForCurrentThreadScope::~DisableMallocRestrictionsForCurrentThreadScope): Deleted.
(WTF::TryMallocReturnValue::TryMallocReturnValue): Deleted.
(WTF::TryMallocReturnValue::~TryMallocReturnValue): Deleted.
(WTF::TryMallocReturnValue::getValue): Deleted.
* Source/WTF/wtf/FixedVector.h:
* Source/WTF/wtf/Forward.h:
* Source/WTF/wtf/MallocCommon.cpp: Copied from Source/JavaScriptCore/jit/JITAllocator.h.
(WTF::ForbidMallocUseForCurrentThreadScope::ForbidMallocUseForCurrentThreadScope):
(WTF::ForbidMallocUseForCurrentThreadScope::~ForbidMallocUseForCurrentThreadScope):
(WTF::DisableMallocRestrictionsForCurrentThreadScope::DisableMallocRestrictionsForCurrentThreadScope):
(WTF::DisableMallocRestrictionsForCurrentThreadScope::~DisableMallocRestrictionsForCurrentThreadScope):
(WTF::assertMallocRestrictionForCurrentThreadScope):
* Source/WTF/wtf/MallocCommon.h: Added.
(WTF::TryMallocReturnValue::TryMallocReturnValue):
(WTF::TryMallocReturnValue::~TryMallocReturnValue):
(WTF::TryMallocReturnValue::getValue):
(WTF::ForbidMallocUseForCurrentThreadScope::~ForbidMallocUseForCurrentThreadScope):
(WTF::DisableMallocRestrictionsForCurrentThreadScope::~DisableMallocRestrictionsForCurrentThreadScope):
(WTF::assertMallocRestrictionForCurrentThreadScope):
* Source/WTF/wtf/MallocSpan.h:
* Source/WTF/wtf/PlatformUse.h:
* Source/WTF/wtf/ProtectedJITMalloc.cpp: Added.
(WTF::sequesteredArenaSetMaxSingleAllocationSize):
(WTF::isSequesteredArenaMallocEnabled):
(WTF::sequesteredArenaMalloc):
(WTF::sequesteredArenaZeroedMalloc):
(WTF::sequesteredArenaCalloc):
(WTF::sequesteredArenaRealloc):
(WTF::sequesteredArenaFree):
(WTF::trySequesteredArenaMalloc):
(WTF::trySequesteredArenaZeroedMalloc):
(WTF::trySequesteredArenaCalloc):
(WTF::trySequesteredArenaRealloc):
(WTF::sequesteredArenaAlignedMalloc):
(WTF::trySequesteredArenaAlignedMalloc):
(WTF::sequesteredArenaAlignedFree):
(WTF::sequesteredArenaMallocStatistics):
(WTF::sequesteredArenaMallocDumpMallocStats):
* Source/WTF/wtf/ProtectedJITMalloc.h: Added.
(WTF::SequesteredArenaAllocator::SequesteredArenaAllocator):
(WTF::SequesteredArenaAllocator::allocate):
(WTF::SequesteredArenaAllocator::deallocate):
(WTF::operator==):
(WTF::SequesteredArenaMalloc::malloc):
(WTF::SequesteredArenaMalloc::tryMalloc):
(WTF::SequesteredArenaMalloc::zeroedMalloc):
(WTF::SequesteredArenaMalloc::tryZeroedMalloc):
(WTF::SequesteredArenaMalloc::realloc):
(WTF::SequesteredArenaMalloc::tryRealloc):
(WTF::SequesteredArenaMalloc::free):
(WTF::SequesteredArenaMalloc::nextCapacity):
(WTF::SequesteredArenaFree::operator() const):
* Source/WTF/wtf/ProtectedJITSpareFile.cpp: Added.
* Source/WTF/wtf/ProtectedJITSpareFile.h: Added.
* Source/WTF/wtf/SegmentedVector.h:
* Source/WTF/wtf/SequesteredAllocator.cpp: Added.
(WTF::SequesteredArenaAllocator::logLiveAllocationDebugInfos):
* Source/WTF/wtf/SequesteredAllocator.h: Added.
(WTF::ArenaLifetime::ArenaLifetime):
(WTF::ArenaLifetime::~ArenaLifetime):
(WTF::ArenaLifetime::isAlive):
* Source/WTF/wtf/SequesteredImmortalHeap.cpp: Added.
* Source/WTF/wtf/SequesteredImmortalHeap.h: Added.
(WTF::SequesteredImmortalHeap::instance):
(WTF::SequesteredImmortalHeap::allocateAndInstall):
(WTF::SequesteredImmortalHeap::getSlot):
(WTF::SequesteredImmortalHeap::computeSlotIndex):
(WTF::SequesteredImmortalHeap::scavenge):
(WTF::SequesteredImmortalHeap::mapPages):
(WTF::SequesteredImmortalHeap::SequesteredImmortalHeap):
(WTF::SequesteredImmortalHeap::getUnchecked):
(WTF::SequesteredImmortalHeap::threadIdForLogging):
* Source/WTF/wtf/StdLibExtras.h:
(WTF::makeUnique):
(WTF::makeUniqueWithoutRefCountedCheck):
* Source/WTF/wtf/ThreadSpecific.h:
(WTF::ThreadSpecific::Data::Data):
(WTF::mustBeCompilerThread&gt;::ThreadSpecific):
(WTF::mustBeCompilerThread&gt;::get):
(WTF::mustBeCompilerThread&gt;::setInTLS):
(WTF::mustBeCompilerThread&gt;::destroy):
(WTF::mustBeCompilerThread&gt;::set):
(WTF::mustBeCompilerThread&gt;::isSet):
(WTF::T):
(WTF::mustBeCompilerThread&gt;::operator):
(WTF::canBeGCThread&gt;::ThreadSpecific): Deleted.
(WTF::canBeGCThread&gt;::get): Deleted.
(WTF::canBeGCThread&gt;::setInTLS): Deleted.
(WTF::canBeGCThread&gt;::destroy): Deleted.
(WTF::canBeGCThread&gt;::set): Deleted.
(WTF::canBeGCThread&gt;::isSet): Deleted.
(WTF::canBeGCThread&gt;::operator): Deleted.
* Source/WTF/wtf/Threading.cpp:
* Source/WTF/wtf/UniqueRef.h:
(WTF::makeUniqueRefWithoutRefCountedCheck):
(WTF::makeUniqueRef):
* Source/WTF/wtf/Vector.h:
* Source/bmalloc/bmalloc/IsoHeap.h:
* Source/bmalloc/bmalloc/IsoHeapInlines.h:
* Source/bmalloc/bmalloc/TZoneHeap.h:
* Source/bmalloc/bmalloc/TZoneHeapInlines.h:

Canonical link: <a href="https://commits.webkit.org/291155@main">https://commits.webkit.org/291155@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/85311d93866d2e52fe9606dec48cd91c408db18c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/92156 "9 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11689 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/1254 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/97060 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/42716 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/12016 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/123/builds/20178 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/97060 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/42716 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/95157 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/12016 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/83468 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/97060 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/12016 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/1082 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/41934 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/84917 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/12016 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/1052 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/99110 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/90868 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/19255 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/123/builds/20178 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/99110 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/19507 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/79331 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/99110 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/798 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/12262 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14640 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/19235 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/24410 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/113477 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/18928 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/32836 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/22386 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/20672 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/WTF/wtf/SequesteredMalloc.h:
```diff
@@ -0,0 +1,244 @@
+
+template<typename T, typename U> inline bool operator==(const SaAllocator<T>&, const SaAllocator<U>&) { return true; }
+
+struct SaMalloc {
```
I like `ArenaMalloc` instead of `Sa`.

Achierius replied:
Renamed to SequesteredArenaMalloc

Constellation reviewed (commented):
sa / Sa are too hard name to read. I like ArenaAllocator / ArenaMalloc etc. instead.

Achierius commented on Source/JavaScriptCore/jit/JIT.cpp:
```diff
@@ -81,7 +82,9 @@ JIT::JIT(VM& vm, BaselineJITPlan& plan, CodeBlock* codeBlock)
-JIT::~JIT() = default;
+JIT::~JIT()
+{
+}
```
Should remove

Constellation commented on Source/WTF/wtf/PlatformUse.h:
```diff
@@ -314,6 +314,14 @@
 #endif
 
+#if !defined(USE_PROTECTED_JIT)
+#if CPU(ARM64) && OS(DARWIN) && (__SIZEOF_POINTER__ == 8)
```
Use `CPU(ADDRESS64)`.

Constellation commented on Source/WTF/wtf/SequesteredAllocator.cpp:
```diff
@@ -0,0 +1,54 @@
+        std::span<const LChar> span = info.proximateFrame.span8();
+        std::string toPrint { span.begin(), span.end() };
+        // No newline since we assume the stack frame will have it
+        dataLogFIf(verbose,
```
Let's use `dataLogLn` and `RawPointer()` instead.
Then you can just pass `info.proximateFrame` directly.

Achierius commented on Source/JavaScriptCore/jit/JIT.h:
```diff
@@ -167,7 +168,6 @@ namespace JSC {
 
     public:
         JIT(VM&, BaselineJITPlan&, CodeBlock*);
-        ~JIT();
```
Unintended change, I'll re-add

Achierius commented on Source/JavaScriptCore/runtime/InitializeThreading.cpp:
```diff
@@ -76,6 +76,7 @@ void initialize()
 
     std::call_once(onceFlag, [] {
         WTF::initialize();
+
```
Unintended change, will remove

Constellation commented on Source/WTF/wtf/SequesteredMalloc.cpp:
```diff
@@ -0,0 +1,158 @@
+
+namespace WTF {
+
+#if !defined(NDEBUG)
```
Use `#if ASSERT_ENABLED`.

Constellation commented on Source/WTF/wtf/SequesteredMalloc.h:
```diff
@@ -0,0 +1,263 @@
+
+namespace WTF {
+
+#if !defined(NDEBUG)
```
Ditto.

Constellation commented on Source/WTF/wtf/SequesteredImmortalHeap.h:
```diff
@@ -0,0 +1,192 @@
+class SequesteredImmortalHeap {
+    static constexpr bool verbose = true;
+    static constexpr pthread_key_t key = __PTK_FRAMEWORK_JAVASCRIPTCORE_KEY0;
+    static constexpr size_t sequesteredImmortalHeapSlotSize { 16 * 1024 };
```
Let's use `16 * KB`.

Constellation commented on Source/WTF/wtf/SequesteredImmortalHeap.h:
```diff
@@ -0,0 +1,192 @@
+#include <wtf/Lock.h>
+
+#if defined(__has_include)
+#if __has_include(<System/pthread_machdep.h>)
```
Usually we should use `OS` macro here. If `pthread_machdep.h` exists on Darwin, let's say `OS(DARWIN)`. If it is Apple-internal SDK, then let's say `USE(APPLE_INTERNAL_SDK)`.

Constellation commented on Source/WTF/wtf/SequesteredAllocator.h:
```diff
@@ -0,0 +1,579 @@
+
+    constexpr static bool verbose { true };
+    constexpr static bool trackAllocationDebugInfo { true };
+    constexpr static size_t pageSize { 1024 * 16 };
```
Use `16 * KB`.

Constellation commented on Source/WTF/wtf/SequesteredAllocator.cpp:
```diff
@@ -0,0 +1,54 @@
+            reinterpret_cast<void*>(p),
+            toPrint.c_str());
+    }
+    dataLogFIf(verbose, "Allocator %d: %zu allocations logged\n", id(), printCount);
```
Let's avoid using F variant as much as possible.
```
dataLogLnIf(verbose, "Allocator ", id(), ": ", printCount, " allocations logged");
```

Constellation commented on Source/WTF/wtf/SequesteredAllocator.h:
```diff
@@ -0,0 +1,579 @@
+
+namespace WTF {
+
+constexpr size_t minArenaGranuleSize { 16 * 1024 };
```
Let's say `16 * KB`.

Constellation commented on Source/WTF/wtf/SequesteredAllocator.h:
```diff
@@ -0,0 +1,579 @@
+        void* allocate(size_t bytes)
+        {
+            auto retval = allocateImpl<AllocationFailureMode::Assert>(bytes);
+            dataLogFIf(verbose,
```
Let's avoid using F variant as much as possible.

Constellation commented on Source/WTF/wtf/SequesteredAllocator.h:
```diff
@@ -0,0 +1,579 @@
+        void* tryAllocate(size_t bytes)
+        {
+            auto retval = allocateImpl<AllocationFailureMode::ReturnNull>(bytes);
+            dataLogFIf(verbose && (retval != nullptr),
```
Ditto.

Constellation commented on Source/WTF/wtf/SequesteredAllocator.h:
```diff
@@ -0,0 +1,579 @@
+        void* alignedAllocate(size_t alignment, size_t bytes)
+        {
+            auto retval = alignedAllocateImpl<AllocationFailureMode::Assert>(alignment, bytes);
+            dataLogFIf(verbose,
```
Ditto.

Constellation commented on Source/WTF/wtf/SequesteredAllocator.h:
```diff
@@ -0,0 +1,579 @@
+        void* tryAlignedAllocate(size_t alignment, size_t bytes)
+        {
+            auto retval = alignedAllocateImpl<AllocationFailureMode::ReturnNull>(alignment, bytes);
+            dataLogFIf(verbose && (retval != nullptr),
```
Ditto.

Constellation commented on Source/WTF/wtf/SequesteredAllocator.h:
```diff
@@ -0,0 +1,579 @@
+            static_assert(sizeof(ArenaGranuleHeader) >= minHeadAlignment);
+            m_allocHead = reinterpret_cast<uintptr_t>(granule) + sizeof(ArenaGranuleHeader);
+            m_allocBound = reinterpret_cast<uintptr_t>(granule) + minArenaGranuleSize;
+            dataLogFIf(verbose,
```
Ditto.

Constellation commented on Source/WTF/wtf/SequesteredAllocator.h:
```diff
@@ -0,0 +1,579 @@
+        ASSERT(!m_alive);
+        m_alive = true;
+        m_liveAllocations = 0;
+        dataLogFIf(verbose,
```
Ditto.

Constellation commented on Source/WTF/wtf/SequesteredAllocator.h:
```diff
@@ -0,0 +1,579 @@
+        m_alive = false;
+
+        WTF::Locker lock(m_decommitLock);
+        dataLogFIf(verbose,
```
Ditto.

Constellation commented on Source/WTF/wtf/SequesteredAllocator.h:
```diff
@@ -0,0 +1,579 @@
+        // TODO: see if we can optimize out the loop by keeping track of the
+        // bottom granule for each arena
+        for (;;) {
+            dataLogFIf(verbose, "    -> %p\n", next);
```
Ditto.

Constellation commented on Source/WTF/wtf/SequesteredAllocator.h:
```diff
@@ -0,0 +1,579 @@
+        if (top == nullptr)
+            return;
+
+        // TODO: see if we can optimize out the loop by keeping track of the
```
Replace TODO with FIXME

Constellation commented on Source/WTF/wtf/SequesteredAllocator.h:
```diff
@@ -0,0 +1,579 @@
+
+    void* reallocateInto(void* from, void* to, size_t toSize)
+    {
+        // TODO: reimplement realloc properly, if necessary
```
Ditto.

Constellation commented on Source/WTF/wtf/SequesteredAllocator.h:
```diff
@@ -0,0 +1,579 @@
+    WTF::Lock m_decommitLock { };
+    ArenaGranuleHeader* m_decommitHead { nullptr };
+    bool m_alive { false };
+    // TODO: wrap this in a child class so we can ifdef it out
```
Ditto.

Constellation commented on Source/WTF/wtf/SequesteredAllocator.h:
```diff
@@ -0,0 +1,579 @@
+    bool m_alive { false };
+    // TODO: wrap this in a child class so we can ifdef it out
+    // when we don't need it
+    std::map<uintptr_t, AllocationDebugInfo> m_allocationInfos;
```
Replace it with StdMap.

Constellation commented on Source/WTF/wtf/SequesteredImmortalHeap.h:
```diff
@@ -0,0 +1,192 @@
+        _pthread_setspecific_direct(key, reinterpret_cast<void*>(slot));
+        pthread_key_init_np(key, nullptr);
+
+        dataLogFIf(verbose, "SequesteredImmortalHeap: thread (%llu) allocated slot %lu (%p)\n", threadIdForLogging(), instance().m_nextFreeIndex - 1, slot);
```
Ditto.

Constellation commented on Source/WTF/wtf/SequesteredAllocator.h:
```diff
@@ -0,0 +1,579 @@
+        return SequesteredImmortalHeap::instance().computeSlotIndex(this);
+    }
+
+    static uint64_t threadIdForLogging()
```
We do not need to have this function in particular.  Let's pass `Thread::current()` to `dataLog`.

Constellation commented on Source/WTF/wtf/SequesteredAllocator.h:
```diff
@@ -0,0 +1,579 @@
+    ALWAYS_INLINE void registerSuccessfulAllocation(void* p, size_t bytes) {
+        m_liveAllocations++;
+
+        (void)bytes;
```
Use `UNUSED_PARAM(bytes)`

Constellation commented on Source/WTF/wtf/SequesteredAllocator.h:
```diff
@@ -0,0 +1,579 @@
+
+    void registerSuccessfulReallocation(void* from, void* to, size_t newSize)
+    {
+        (void)newSize;
```
Ditto.

Constellation commented on Source/WTF/wtf/SequesteredImmortalHeap.h:
```diff
@@ -0,0 +1,192 @@
+        auto slot = reinterpret_cast<uintptr_t>(slotPtr);
+        auto arrayBase = reinterpret_cast<uintptr_t>(m_slots.begin());
+        [[maybe_unused]] auto arrayBound = reinterpret_cast<uintptr_t>(m_slots.begin()) + sizeof(m_slots);
+        ASSERT(slot >= arrayBase && slot < arrayBound);
```
Use `ASSERT_UNUSED(arrayBound, ...)` and drop `[[maybe_unused]]`.

Constellation commented on Source/WTF/wtf/SequesteredImmortalHeap.h:
```diff
@@ -0,0 +1,192 @@
+
+    static void scavenge()
+    {
+        // TODO: provide hook for libpas scavenger
```
WebKit uses FIXME

Constellation commented on Source/WTF/wtf/SequesteredMalloc.cpp:
```diff
@@ -0,0 +1,158 @@
+void* sequesteredArenaMalloc(size_t size)
+{
+    ASSERT_IS_WITHIN_LIMIT(size);
+    return WTF::SequesteredArenaAllocator::getCurrentAllocator()->malloc(size);
```
Drop WTF::

Constellation commented on Source/WTF/wtf/SequesteredMalloc.cpp:
```diff
@@ -0,0 +1,158 @@
+void* sequesteredArenaZeroedMalloc(size_t size)
+{
+    ASSERT_IS_WITHIN_LIMIT(size);
+    return WTF::SequesteredArenaAllocator::getCurrentAllocator()->zeroedMalloc(size);
```
Drop WTF::

Constellation commented on Source/WTF/wtf/SequesteredMalloc.cpp:
```diff
@@ -0,0 +1,158 @@
+void* sequesteredArenaRealloc(void* object, size_t newSize)
+{
+    ASSERT_IS_WITHIN_LIMIT(newSize);
+    return WTF::SequesteredArenaAllocator::getCurrentAllocator()->realloc(object, newSize);
```
Ditto.

Constellation commented on Source/WTF/wtf/SequesteredMalloc.cpp:
```diff
@@ -0,0 +1,158 @@
+
+void sequesteredArenaFree(void* object)
+{
+    WTF::SequesteredArenaAllocator::getCurrentAllocator()->free(object);
```
Ditto.

Constellation commented on Source/WTF/wtf/SequesteredMalloc.cpp:
```diff
@@ -0,0 +1,158 @@
+TryMallocReturnValue trySequesteredArenaMalloc(size_t size)
+{
+    FAIL_IF_EXCEEDS_LIMIT(size);
+    return WTF::SequesteredArenaAllocator::getCurrentAllocator()->tryMalloc(size);
```
Ditto.

Constellation commented on Source/WTF/wtf/SequesteredMalloc.cpp:
```diff
@@ -0,0 +1,158 @@
+TryMallocReturnValue trySequesteredArenaZeroedMalloc(size_t size)
+{
+    FAIL_IF_EXCEEDS_LIMIT(size);
+    return WTF::SequesteredArenaAllocator::getCurrentAllocator()->tryZeroedMalloc(size);
```
Ditto.

Constellation commented on Source/WTF/wtf/SequesteredMalloc.cpp:
```diff
@@ -0,0 +1,158 @@
+TryMallocReturnValue trySequesteredArenaRealloc(void* object, size_t newSize)
+{
+    FAIL_IF_EXCEEDS_LIMIT(newSize);
+    return WTF::SequesteredArenaAllocator::getCurrentAllocator()->tryRealloc(object, newSize);
```
Ditto.

Constellation commented on Source/WTF/wtf/SequesteredMalloc.cpp:
```diff
@@ -0,0 +1,158 @@
+void* sequesteredArenaAlignedMalloc(size_t alignment, size_t size)
+{
+    ASSERT_IS_WITHIN_LIMIT(size);
+    return WTF::SequesteredArenaAllocator::getCurrentAllocator()->alignedMalloc(alignment, size);
```
Ditto.

Constellation commented on Source/WTF/wtf/SequesteredMalloc.cpp:
```diff
@@ -0,0 +1,158 @@
+void* trySequesteredArenaAlignedMalloc(size_t alignment, size_t size)
+{
+    FAIL_IF_EXCEEDS_LIMIT(size);
+    return WTF::SequesteredArenaAllocator::getCurrentAllocator()->tryAlignedMalloc(alignment, size);
```
Ditto.

Constellation commented on Source/WTF/wtf/SequesteredMalloc.cpp:
```diff
@@ -0,0 +1,158 @@
+}
+void sequesteredArenaAlignedFree(void* p)
+{
+    return WTF::SequesteredArenaAllocator::getCurrentAllocator()->free(p);
```
Ditto.

Constellation commented on Source/WTF/wtf/SequesteredImmortalHeap.h:
```diff
@@ -0,0 +1,192 @@
+    {
+        void* memory = mmap(nullptr, bytes, PROT_READ | PROT_WRITE,
+            MAP_PRIVATE | MAP_ANON, -1, 0);
+        if (memory == MAP_FAILED) {
```
Put `UNLIKELY`.

Constellation commented on Source/WTF/wtf/SequesteredImmortalHeap.h:
```diff
@@ -0,0 +1,192 @@
+    void* getSlot()
+    {
+        auto* val = getUnchecked();
+        return val;
```
Let's just make it `return getUnchecked();`

Constellation commented on Source/WTF/wtf/SequesteredImmortalHeap.h:
```diff
@@ -0,0 +1,192 @@
+        {
+            WTF::Locker locker { m_scavengerLock };
+            ASSERT(getUnchecked() == nullptr);
+            // TODO: implement resizing to a larger capacity
```
Use FIXME

Constellation commented on Source/WTF/wtf/SequesteredImmortalHeap.h:
```diff
@@ -0,0 +1,192 @@
+            // TODO: implement resizing to a larger capacity
+            RELEASE_ASSERT(m_nextFreeIndex < numSlots);
+
+            IGNORE_WARNINGS_BEGIN("unsafe-buffer-usage")
```
Why not use WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN?

Constellation commented on Source/WTF/wtf/SequesteredImmortalHeap.h:
```diff
@@ -0,0 +1,192 @@
+            IGNORE_WARNINGS_BEGIN("unsafe-buffer-usage")
+            void* buff = &(m_slots[m_nextFreeIndex++]);
+            slot = new (buff) T();
+            IGNORE_WARNINGS_END
```
Ditto.

Constellation commented on Source/WTF/wtf/SequesteredAllocator.h:
```diff
@@ -0,0 +1,579 @@
+            dataLogFIf(verbose, "    -> %p\n", next);
+            auto& header = ArenaGranuleHeader::headerForGranule(next);
+            if (header.priorGranule == nullptr) {
+                dataLogFIf(verbose, "    -> done\n");
```
Ditto.

Constellation commented on Source/WTF/wtf/SequesteredAllocator.h:
```diff
@@ -0,0 +1,579 @@
+        {
+            void* memory = mapPages<mode>(minArenaGranuleSize);
+            if constexpr (mode == AllocationFailureMode::ReturnNull) {
+                if (memory == nullptr)
```
Let's use `!memory` for nullptr check in general. Ditto for the rest of the code.

Constellation reviewed (commented):
Commented.

Constellation commented on Source/WTF/wtf/SequesteredAllocator.h:
```diff
@@ -0,0 +1,579 @@
+        {
+            uintptr_t allocation = headAlignedUpTo(alignment);
+            uintptr_t newHead = headIncrementedBy((allocation - m_allocHead) + bytes);
+            if (newHead < m_allocBound) {
```
Put `LIKELY`.

Constellation commented on Source/WTF/wtf/SequesteredAllocator.h:
```diff
@@ -0,0 +1,579 @@
+        {
+            uintptr_t allocation = m_allocHead;
+            uintptr_t newHead = headIncrementedBy(bytes);
+            if (newHead < m_allocBound) {
```
Put `LIKELY`.

Constellation commented on Source/WTF/wtf/SequesteredImmortalHeap.h:
```diff
@@ -0,0 +1,192 @@
+namespace WTF {
+
+class SequesteredImmortalHeap {
+    static constexpr bool verbose = true;
```
When landing, `verbose` etc. should be `false`.

Constellation commented on Source/WTF/wtf/SequesteredImmortalHeap.h:
```diff
@@ -0,0 +1,192 @@
+        return _pthread_getspecific_direct(key);
+    }
+
+    static uint64_t threadIdForLogging()
```
Use `Thread::current()`

Constellation commented on Source/WTF/wtf/SequesteredAllocator.h:
```diff
@@ -0,0 +1,579 @@
+    {
+        ASSERT(m_alive);
+        auto retval = m_genericSmallArena.tryAllocate(bytes);
+        if (retval != nullptr)
```
We should put LIKELY. And ditto for similar places.

Constellation commented on Source/WTF/wtf/SequesteredImmortalHeap.h:
```diff
@@ -0,0 +1,192 @@
+    {
+        auto ptr = reinterpret_cast<SequesteredImmortalHeap*>(&s_instance);
+        if (UNLIKELY(!ptr->m_initialized))
+            new (&s_instance) SequesteredImmortalHeap();
```
You need to use `std::call_once`
And `m_initialized` initialization order is not well enforced (no store barrier). So let's not do this hand-made singleton and use LazyNeverDestroyed & std::call_once simply.

Achierius replied:
I can't use LazyNeverDestroyed since its initialization flag would lie outside of the sequestered memory region. I'll use std::once_flag for now but we'll need to revisit it later

Achierius committed:
[JSC] Implement core of sequestered arena allocator
https://bugs.webkit.org/show_bug.cgi?id=287524
rdar://114840482

Reviewed by Yusuke Suzuki.

This implements the core functionality of a new arena allocator built
around a sequestered pool of memory. The intention is for this allocator
to be used by the JIT compiler for allocating datastructures; as of
right now, only a small number of datatypes are set up to use the new
allocator. As this is an arena allocator, users need to specify the
lifetime of their allocations indirectly via the ArenaLifetime object.
Currently, the allocator is disabled, and has been stripped of several
unstable components which are still under development. In particular,
the `free` flow is intended to hook into the libpas scavenger, but right
now eagerly frees to tide us over; moreover, each thread's allocator
will eventually have at least a small number (2-3) of different
size-classed arenas, but for now we only use one.

This allocator is by design separate from libpas. This is because we
want to isolate not only the backing memory but also all allocator
metadata and internal memory. To do this within libpas would require we
duplicate essentially the entire libpas heap hierarchy, from
pas_large_heap up to the various bootstrap heaps, and color them
according to which pool of pages they're supposed to pull from.

Performance-wise, turning this on has no effect at the moment, but the
hope is that in time it will prove to be a progression. The arena
semantics allow us to batch all frees from a given compiler pass into
one, reducing the actual cost of `free` to a no-op as well as reducing
the load on the scavenger. It also allows us to significantly simplify
the allocation path, removing some overhead even in the shortest-path
malloc case and making it possible to stick to the bump-allocation
regime without ever having to fall back on free lists or &c.

In time, we will need to add a persistent allocator which allocates from
the same pool as well -- something like SEQUESTERED_PERSISTENT_MALLOC.

With regards to the naming, I considered a couple of alternatives:
  - Protected JIT Arena Malloc
  - JIT Arena Malloc
  - Secure Arena Malloc
For the first two, I want to avoid using the term "JIT" as it's very
overloaded -- I've seen it refer to things relevant to
  A) The JIT compiler
  B) Allocating RWX memory
  C) Threads executing in previously JITted code
and so would like to avoid using it here. It also has the problem that
the short-form (JaMalloc) sounds a lot like jemalloc, which could cause
further confusion.
"Secure" would imply that libpas/FastMalloc are "insecure", which I also
want to avoid.
"Sequestered" is nice because it describes exactly what characterizes
this allocator: it uses memory sequestered from the rest of the system,
and it allocates it using an arena.
Thus SequesteredArenaMalloc.

* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/jit/JIT.cpp:
(JSC::JIT::~JIT):
* Source/JavaScriptCore/jit/JIT.h:
* Source/JavaScriptCore/jit/JITAllocator.h:
* Source/JavaScriptCore/jit/JITWorklistThread.cpp:
* Source/JavaScriptCore/jit/SecureAllocator.cpp: Copied from Source/JavaScriptCore/jit/JITAllocator.h.
* Source/JavaScriptCore/jit/SecureAllocator.h: Copied from Source/JavaScriptCore/jit/JITAllocator.h.
* Source/JavaScriptCore/jit/SecureAllocatorInlines.h: Copied from Source/JavaScriptCore/jit/JITAllocator.h.
* Source/JavaScriptCore/runtime/InitializeThreading.cpp:
(JSC::initialize):
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/Bag.h:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/CheckedRef.h:
(WTF::CanMakeCheckedPtr::~CanMakeCheckedPtr):
(WTF::CanMakeThreadSafeCheckedPtr::~CanMakeThreadSafeCheckedPtr):
* Source/WTF/wtf/EmbeddedFixedVector.h:
* Source/WTF/wtf/FastMalloc.cpp:
(WTF::tryFastMalloc):
(WTF::fastMalloc):
(WTF::tryFastCalloc):
(WTF::fastCalloc):
(WTF::fastRealloc):
(WTF::tryFastRealloc):
(WTF::fastZeroedMalloc):
(WTF::tryFastZeroedMalloc):
(WTF::fastAlignedMalloc):
(WTF::tryFastAlignedMalloc):
(WTF::fastCompactMalloc):
(WTF::fastCompactZeroedMalloc):
(WTF::tryFastCompactZeroedMalloc):
(WTF::fastCompactRealloc):
(WTF::fastCompactAlignedMalloc):
(WTF::tryFastCompactAlignedMalloc):
(WTF::tryFastCompactMalloc):
(WTF::tryFastCompactRealloc):
(WTF::ForbidMallocUseForCurrentThreadScope::ForbidMallocUseForCurrentThreadScope): Deleted.
(WTF::ForbidMallocUseForCurrentThreadScope::~ForbidMallocUseForCurrentThreadScope): Deleted.
(WTF::DisableMallocRestrictionsForCurrentThreadScope::DisableMallocRestrictionsForCurrentThreadScope): Deleted.
(WTF::DisableMallocRestrictionsForCurrentThreadScope::~DisableMallocRestrictionsForCurrentThreadScope): Deleted.
* Source/WTF/wtf/FastMalloc.h:
(WTF::FastMalloc::fastFree):
(WTF::FastCompactMalloc::fastFree):
(WTF::ForbidMallocUseForCurrentThreadScope::~ForbidMallocUseForCurrentThreadScope): Deleted.
(WTF::DisableMallocRestrictionsForCurrentThreadScope::~DisableMallocRestrictionsForCurrentThreadScope): Deleted.
(WTF::TryMallocReturnValue::TryMallocReturnValue): Deleted.
(WTF::TryMallocReturnValue::~TryMallocReturnValue): Deleted.
(WTF::TryMallocReturnValue::getValue): Deleted.
* Source/WTF/wtf/FixedVector.h:
* Source/WTF/wtf/Forward.h:
* Source/WTF/wtf/MallocCommon.cpp: Copied from Source/JavaScriptCore/jit/JITAllocator.h.
(WTF::ForbidMallocUseForCurrentThreadScope::ForbidMallocUseForCurrentThreadScope):
(WTF::ForbidMallocUseForCurrentThreadScope::~ForbidMallocUseForCurrentThreadScope):
(WTF::DisableMallocRestrictionsForCurrentThreadScope::DisableMallocRestrictionsForCurrentThreadScope):
(WTF::DisableMallocRestrictionsForCurrentThreadScope::~DisableMallocRestrictionsForCurrentThreadScope):
(WTF::assertMallocRestrictionForCurrentThreadScope):
* Source/WTF/wtf/MallocCommon.h: Added.
(WTF::TryMallocReturnValue::TryMallocReturnValue):
(WTF::TryMallocReturnValue::~TryMallocReturnValue):
(WTF::TryMallocReturnValue::getValue):
(WTF::ForbidMallocUseForCurrentThreadScope::~ForbidMallocUseForCurrentThreadScope):
(WTF::DisableMallocRestrictionsForCurrentThreadScope::~DisableMallocRestrictionsForCurrentThreadScope):
(WTF::assertMallocRestrictionForCurrentThreadScope):
* Source/WTF/wtf/MallocSpan.h:
* Source/WTF/wtf/PlatformUse.h:
* Source/WTF/wtf/ProtectedJITMalloc.cpp: Added.
(WTF::sequesteredArenaSetMaxSingleAllocationSize):
(WTF::isSequesteredArenaMallocEnabled):
(WTF::sequesteredArenaMalloc):
(WTF::sequesteredArenaZeroedMalloc):
(WTF::sequesteredArenaCalloc):
(WTF::sequesteredArenaRealloc):
(WTF::sequesteredArenaFree):
(WTF::trySequesteredArenaMalloc):
(WTF::trySequesteredArenaZeroedMalloc):
(WTF::trySequesteredArenaCalloc):
(WTF::trySequesteredArenaRealloc):
(WTF::sequesteredArenaAlignedMalloc):
(WTF::trySequesteredArenaAlignedMalloc):
(WTF::sequesteredArenaAlignedFree):
(WTF::sequesteredArenaMallocStatistics):
(WTF::sequesteredArenaMallocDumpMallocStats):
* Source/WTF/wtf/ProtectedJITMalloc.h: Added.
(WTF::SequesteredArenaAllocator::SequesteredArenaAllocator):
(WTF::SequesteredArenaAllocator::allocate):
(WTF::SequesteredArenaAllocator::deallocate):
(WTF::operator==):
(WTF::SequesteredArenaMalloc::malloc):
(WTF::SequesteredArenaMalloc::tryMalloc):
(WTF::SequesteredArenaMalloc::zeroedMalloc):
(WTF::SequesteredArenaMalloc::tryZeroedMalloc):
(WTF::SequesteredArenaMalloc::realloc):
(WTF::SequesteredArenaMalloc::tryRealloc):
(WTF::SequesteredArenaMalloc::free):
(WTF::SequesteredArenaMalloc::nextCapacity):
(WTF::SequesteredArenaFree::operator() const):
* Source/WTF/wtf/ProtectedJITSpareFile.cpp: Added.
* Source/WTF/wtf/ProtectedJITSpareFile.h: Added.
* Source/WTF/wtf/SegmentedVector.h:
* Source/WTF/wtf/SequesteredAllocator.cpp: Added.
(WTF::SequesteredArenaAllocator::logLiveAllocationDebugInfos):
* Source/WTF/wtf/SequesteredAllocator.h: Added.
(WTF::ArenaLifetime::ArenaLifetime):
(WTF::ArenaLifetime::~ArenaLifetime):
(WTF::ArenaLifetime::isAlive):
* Source/WTF/wtf/SequesteredImmortalHeap.cpp: Added.
* Source/WTF/wtf/SequesteredImmortalHeap.h: Added.
(WTF::SequesteredImmortalHeap::instance):
(WTF::SequesteredImmortalHeap::allocateAndInstall):
(WTF::SequesteredImmortalHeap::getSlot):
(WTF::SequesteredImmortalHeap::computeSlotIndex):
(WTF::SequesteredImmortalHeap::scavenge):
(WTF::SequesteredImmortalHeap::mapPages):
(WTF::SequesteredImmortalHeap::SequesteredImmortalHeap):
(WTF::SequesteredImmortalHeap::getUnchecked):
(WTF::SequesteredImmortalHeap::threadIdForLogging):
* Source/WTF/wtf/StdLibExtras.h:
(WTF::makeUnique):
(WTF::makeUniqueWithoutRefCountedCheck):
* Source/WTF/wtf/ThreadSpecific.h:
(WTF::ThreadSpecific::Data::Data):
(WTF::mustBeCompilerThread>::ThreadSpecific):
(WTF::mustBeCompilerThread>::get):
(WTF::mustBeCompilerThread>::setInTLS):
(WTF::mustBeCompilerThread>::destroy):
(WTF::mustBeCompilerThread>::set):
(WTF::mustBeCompilerThread>::isSet):
(WTF::T):
(WTF::mustBeCompilerThread>::operator):
(WTF::canBeGCThread>::ThreadSpecific): Deleted.
(WTF::canBeGCThread>::get): Deleted.
(WTF::canBeGCThread>::setInTLS): Deleted.
(WTF::canBeGCThread>::destroy): Deleted.
(WTF::canBeGCThread>::set): Deleted.
(WTF::canBeGCThread>::isSet): Deleted.
(WTF::canBeGCThread>::operator): Deleted.
* Source/WTF/wtf/Threading.cpp:
* Source/WTF/wtf/UniqueRef.h:
(WTF::makeUniqueRefWithoutRefCountedCheck):
(WTF::makeUniqueRef):
* Source/WTF/wtf/Vector.h:
* Source/bmalloc/bmalloc/IsoHeap.h:
* Source/bmalloc/bmalloc/IsoHeapInlines.h:
* Source/bmalloc/bmalloc/TZoneHeap.h:
* Source/bmalloc/bmalloc/TZoneHeapInlines.h:

Canonical link: https://commits.webkit.org/291155@main
